dlib/media/ffmpeg_abstract.h

```diff
@@ -0,0 +1,955 @@
+                ensures
+                    - returns channel_layout of audio frames to be returned by read()
+                    - If decoder_audio_args::channel_layout > 0, then frames returned by read() are automatically resampled and converted.
+                    - See documentation of AVFrame::channel_layout
+            !*/
+
+            AVSampleFormat sample_fmt() const noexcept;
+            /*!
+                requires
+                    - is_audio_decoder() == true
+                    - must have called push_encoded() enough times such that a call to read() would have returned a frame.
+                ensures
+                    - returns sample format (s16, u8, etc) of audio frames to be returned by read()
+                    - If decoder_audio_args::fmt > 0, then frames returned by read() are automatically resampled and converted.
+            !*/
+
+            int nchannels() const noexcept;
+            /*!
+                requires
+                    - is_audio_decoder() == true
+                    - must have called push_encoded() enough times such that a call to read() would have returned a frame.
+                ensures
+                    - returns number of channels of audio frames to be returned by read()
+                    - If decoder_audio_args::channel_layout > 0, then frames returned by read() are automatically resampled and converted.
+                    - See documentation of AVFrame::channel_layout
+            !*/
+
+            bool push_encoded(const uint8_t *encoded, int nencoded);
+            /*!
+                requires
+                    - is_open() == true
+                ensures
+                    - encodes data.
+                    - if (returns true)
+                        return value of nframes_available() might change unless more input is required
+                    else
+                        decoder is closed but frames may still be available when calling read() so is_open() may still return true
+            !*/
+
+            void flush();
+            /*!
+                requires
+                    - is_open() == true
+                ensures
+                    - Flushes the decoder. This must be called when there is no more data to be decoded. Last remaining frames will be available.
+                    - calls push_encoded(nullptr, 0)
+            !*/
+
+            decoder_status read(frame& dst_frame);
+            /*!
+                ensures
+                    - if (DECODER_EAGAIN)
+                        dst_frame.is_empty() == true
+                        need to call push_encoded() again or flush()
+                    else if (DECODER_FRAME_AVAILABLE)
+                        dst_frame.is_empty() == false
+                    else if (DECODER_CLOSED)
+                        is_open() == false
+                        dst_frame.is_empty() == true
```

Comment:
```suggestion
            /*!
                ensures
                    - Attempts to read a frame, storing the result in dst_frame.
                    - If it is successful then returns DECODER_FRAME_AVAILABLE and it
                      will dst_frame.is_empty() == false.  Otherwise, returns one of the 
                      following:
                         - DECODER_EAGAIN: this indicates (IDK, add some words here).  In particular,
                           you will need to call push_encoded() again or flush() before calling this
                           function (I'm guessing, is this right?  Add more docs about this case to
                           make it clear).
                        - DECODER_CLOSED: This indicates there aren't any more frames.  If this happens
                          then it also means that is_open() == false.

```
Something like that.  Except fill out the parts where I was like "IDK, what does this do?" :)